src/data/num/prime.lean

```diff
@@ -0,0 +1,99 @@
+/-
+Copyright (c) 2020 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Mario Carneiro
+-/
+import data.num.lemmas
+import data.nat.prime
+import tactic.ring
+
+/-!
+# Primality for binary natural numbers
+
+Decision procedures for primality on `num`.
+-/
+
+namespace pos_num
+
+/-- Auxiliary function for computing the smallest prime factor of a `pos_num`. Unlike
+`nat.min_fac_aux`, we use a natural number `fuel` variable that is set to an upper bound on the
+number of iterations. It is initialized to the number `n` we are determining primality for. Even
+though this is exponential in the input (since it is a `nat`, not a `num`), it will get lazily
+evaluated during kernel reduction, so we will only require about `sqrt n` unfoldings, for the
+`sqrt n` iterations of the loop. -/
+def min_fac_aux (n : pos_num) : ℕ → pos_num → pos_num
+| 0 _ := n
+| (fuel+1) k :=
+  if h : n < k.bit1 * k.bit1 then n else
+  if k.bit1 ∣ n then k.bit1 else
+  min_fac_aux fuel k.succ
+
+theorem min_fac_aux_to_nat {fuel:ℕ} {n k : pos_num} (h : nat.sqrt n < fuel + k.bit1) :
+  (min_fac_aux n fuel k : ℕ) = nat.min_fac_aux n k.bit1 :=
+begin
+  induction fuel with fuel ih generalizing k; rw [min_fac_aux, nat.min_fac_aux],
+  { rw if_pos, rwa [zero_add, nat.sqrt_lt] at h },
+  rw [← mul_to_nat], simp only [cast_lt, dvd_to_nat, ite_cast],
+  congr' 2,
+  rw ih; [congr, convert nat.lt_succ_of_lt h using 1];
+  simp only [_root_.bit1, _root_.bit0, cast_bit1, cast_succ,
+    nat.succ_eq_add_one, add_assoc, add_left_comm]
+end
+
+/-- Returns the smallest prime factor of `n ≠ 1`. -/
+def min_fac : pos_num → pos_num
+| 1 := 1
+| (bit0 n) := 2
+| (bit1 n) := min_fac_aux (bit1 n) n 1
+
+@[simp] theorem min_fac_to_nat (n : pos_num) : (min_fac n : ℕ) = nat.min_fac n :=
+begin
+  cases n, {refl},
+  { rw [min_fac, nat.min_fac_eq, if_neg], swap, {simp},
+    rw [min_fac_aux_to_nat], {refl},
+    simp only [cast_one, cast_bit1],
+    rw [nat.sqrt_lt],
+    convert lt_add_of_pos_right _ (dec_trivial : (0:ℕ) < (n+4)*n + 8),
+    unfold _root_.bit1 _root_.bit0, ring },
+  { rw [min_fac, nat.min_fac_eq, if_pos], {refl},
```

Comment:
```suggestion
  cases n, { refl },
  { rw [min_fac, nat.min_fac_eq, if_neg], swap, { simp },
    rw [min_fac_aux_to_nat], { refl },
    simp only [cast_one, cast_bit1],
    rw [nat.sqrt_lt],
    convert lt_add_of_pos_right _ (dec_trivial : (0:ℕ) < (n+4)*n + 8),
    unfold _root_.bit1 _root_.bit0, ring },
  { rw [min_fac, nat.min_fac_eq, if_pos], { refl },
```

Reply:
I don't usually use spaces around end-of-line side goal proofs like this. Has the style guide changed?

Reply:
There's spaces around `simpa` [here in the style guide](https://leanprover-community.github.io/contribute/style.html#tactic-mode). I thought we always put spaces whenever we used braces around tactics, but maybe things were different before.

Reply:
It's not a perfect poll, but I found 35 hits for `, { refl }` and 89 hits for `, {refl}` in mathlib. For `, {simpa` vs `, { simpa` it's about equal numbers. I prefer the style used here (especially for one word tactic proofs) so I think we should keep it as is in the PR and maybe consider a style change PR later across the whole library if we want to recommend one over the other.

Reply:
OK, sounds reasonable!